pkg/function/restore_csi_snapshot.go

```diff
@@ -0,0 +1,162 @@
+	_ = kanister.Register(&restoreCSISnapshotFunc{})
+}
+
+var (
+	_ kanister.Func = (*restoreCSISnapshotFunc)(nil)
+)
+
+const (
+	// Snapshot API Group
+	SnapshotAPIGroup = "snapshot.storage.k8s.io"
+	// RestoreCSISnapshotFuncName gives the name of the function
+	RestoreCSISnapshotFuncName = "RestoreCSISnapshot"
+	// RestoreCSISnapshotNameArg provides name of the VolumeSnapshot
+	RestoreCSISnapshotNameArg = "name"
+	// RestoreCSISnapshotPVCNameArg gives the name of the newly restored PVC
+	RestoreCSISnapshotPVCNameArg = "pvc"
+	// RestoreCSISnapshotNamespaceArg mentions the namespace of the newly restored PVC
+	RestoreCSISnapshotNamespaceArg = "namespace"
+	// RestoreCSISnapshotStorageClassArg specifies the name of the StorageClass
+	RestoreCSISnapshotStorageClassArg = "storageClass"
+	// RestoreCSISnapshotStorageRequestArg provides the storage size to be requested for PV in PVC
+	RestoreCSISnapshotRestoreSizeArg = "restoreSize"
+	// RestoreCSISnapshotAccessModesArg lists down the accessmodes for the underlying PV
+	RestoreCSISnapshotAccessModesArg = "accessModes"
+	// RestoreCSISnapshotLabelsArg has labels that will be added to the newly restored PVC
+	RestoreCSISnapshotLabelsArg = "labels"
+	// RestoreCSISnapshotVolumeModeArg defines mode of volume
+	RestoreCSISnapshotVolumeModeArg = "volumeMode"
+)
+
+type restoreCSISnapshotFunc struct{}
+
+type restoreCSISnapshotArgs struct {
+	Name         string
+	PVC          string
+	Namespace    string
+	StorageClass string
+	RestoreSize  *resource.Quantity
+	AccessModes  []v1.PersistentVolumeAccessMode
+	Labels       map[string]string
+	VolumeMode   v1.PersistentVolumeMode
+}
+
+func (*restoreCSISnapshotFunc) Name() string {
+	return RestoreCSISnapshotFuncName
+}
+
+func (*restoreCSISnapshotFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var restoreSize string
+	var restoreArgs restoreCSISnapshotArgs
+	if err := Arg(args, RestoreCSISnapshotNameArg, &restoreArgs.Name); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotPVCNameArg, &restoreArgs.PVC); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotNamespaceArg, &restoreArgs.Namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotStorageClassArg, &restoreArgs.StorageClass); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotRestoreSizeArg, &restoreSize); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotAccessModesArg, &restoreArgs.AccessModes, []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotVolumeModeArg, &restoreArgs.VolumeMode, v1.PersistentVolumeFilesystem); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotLabelsArg, &restoreArgs.Labels, nil); err != nil {
+		return nil, err
+	}
+	size := resource.MustParse(restoreSize)
```

Comment:
This can crash. We use `ParseQuantity` and handle error instead
https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#ParseQuantity